src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs

```diff
@@ -0,0 +1,260 @@
+
+            string assemblyName = GetAssemblyName(metadataReader).FullName;
+
+            bool isAssemblyComVisible = IsComVisible(metadataReader, metadataReader.GetAssemblyDefinition());
+
+            foreach (TypeDefinitionHandle type in metadataReader.TypeDefinitions)
+            {
+                TypeDefinition definition = metadataReader.GetTypeDefinition(type);
+
+                if (TypeIsPublic(metadataReader, definition) && TypeIsClass(metadataReader, definition) && IsComVisible(metadataReader, definition, isAssemblyComVisible))
+                {
+                    string guid = GetTypeGuid(metadataReader, definition);
+
+                    if (clsidMap.ContainsKey(guid))
+                    {
+                        throw new BuildErrorException(Strings.ClsidMapConflictingGuids, clsidMap[guid].type, GetTypeName(metadataReader, definition), guid);
+                    }
+
+                    clsidMap.Add(guid,
+                        new ClsidEntry
+                        {
+                            type = GetTypeName(metadataReader, definition),
+                            assembly = assemblyName
+                        });
+                }
+            }
+
+            using (StreamWriter writer = File.CreateText(clsidMapPath))
+            {
+                writer.Write(JsonConvert.SerializeObject(clsidMap));
+            }
+        }
+
+        private static bool TypeIsClass(MetadataReader metadataReader, TypeDefinition definition)
+        {
+            if ((definition.Attributes & TypeAttributes.Interface) != 0)
+            {
+                return false;
+            }
+
+            EntityHandle baseTypeEntity = definition.BaseType;
+            if (baseTypeEntity.Kind == HandleKind.TypeReference)
+            {
+                TypeReference baseClass = metadataReader.GetTypeReference((TypeReferenceHandle)baseTypeEntity);
+                if (baseClass.ResolutionScope.Kind == HandleKind.AssemblyReference && ReferenceToCoreLibrary(metadataReader, (AssemblyReferenceHandle)baseClass.ResolutionScope))
+                {
+                    string baseClassName = GetTypeName(metadataReader, baseClass);
+                    if (baseClassName == "System.ValueType" || baseClassName == "System.Enum")
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static readonly string[] s_CoreLibraryNames = { "mscorlib", "System.Runtime", "netstandard" };
+
+        private static bool ReferenceToCoreLibrary(MetadataReader reader, AssemblyReferenceHandle assembly)
+        {
+            return s_CoreLibraryNames.Contains(reader.GetString(reader.GetAssemblyReference(assembly).Name));
+        }
+
+        private static bool TypeIsPublic(MetadataReader reader, TypeDefinition type)
+        {
+            if ((type.Attributes & TypeAttributes.Public) != 0)
```

Comment:
Watch out here. this "flags" enum is a pit of failure.

The correct way to check visibility is to to mask with TypeAttributes.VisibilityMask (0x7) and see which of the eight values you get.

Above will be true for every odd visibility:  e.g. `(NestedPrivate=0x3) & (Public=0x1)` != 0)